src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,763 @@
+//
+//            if constexpr(std::is_integral_v<OffsetDataType> && std::is_integral_v<CharCountDataType>) {
+//              result = evaluate_substring(
+//                evaluate_expression<std::string>(*expression.arguments[0]),
+//                evaluate_expression<OffsetDataType>(*expression.arguments[1]),
+//                evaluate_expression<OffsetDataType>(*expression.arguments[2]));
+//            } else {
+//              Fail("SUBSTRING parameters 2 and 3 need to be integral");
+//            }
+//          });
+//        });
+//      } else {
+//        Fail("SUBSTRING can only return String");
+//      }
+//      // clang-format on
+//  }
+//
+//  return result;
+//}
+
+//template<typename OffsetDataType, typename CharCountDataType>
+//ExpressionResult<std::string> ExpressionEvaluator::evaluate_substring(const ExpressionResult<std::string>& string_result,
+//                                                 const ExpressionResult<OffsetDataType>& offset_result,
+//                                                 const ExpressionResult<CharCountDataType>& char_count_result) {
+//  Fail("Not yet implemented");
+////  /**
+////   *
+////   */
+////
+////  auto result_values = std::vector<std::string>(_output_row_count);
+////  auto result_nulls = std::vector<bool>(_output_row_count);
+////
+////  for (auto chunk_offset = 0; chunk_offset < _output_row_count; ++chunk_offset) {
+////
+////  }
+//}
+
+template<typename T>
+ExpressionResult<T> ExpressionEvaluator::evaluate_arithmetic_expression(const ArithmeticExpression& expression) {
+  const auto& left = *expression.left_operand();
+  const auto& right = *expression.right_operand();
+
+  // clang-format off
+  switch (expression.arithmetic_operator) {
+    case ArithmeticOperator::Addition:       return evaluate_binary_with_default_null_logic<T, Addition>(left, right);
+    case ArithmeticOperator::Subtraction:    return evaluate_binary_with_default_null_logic<T, Subtraction>(left, right);
+    case ArithmeticOperator::Multiplication: return evaluate_binary_with_default_null_logic<T, Multiplication>(left, right);
+    case ArithmeticOperator::Division:       return evaluate_binary_with_default_null_logic<T, Division>(left, right);
+
+    default:
+      Fail("ArithmeticOperator evaluation not yet implemented");
+  }
+  // clang-format on
+}
+
+template<>
+ExpressionResult<int32_t> ExpressionEvaluator::evaluate_binary_predicate_expression<int32_t>(const BinaryPredicateExpression& expression) {
+  const auto& left = *expression.left_operand();
+  const auto& right = *expression.right_operand();
+
+  // clang-format off
+  switch (expression.predicate_condition) {
+    case PredicateCondition::Equals:            return evaluate_binary_with_default_null_logic<int32_t, Equals>(left, right);
+    case PredicateCondition::NotEquals:         return evaluate_binary_with_default_null_logic<int32_t, NotEquals>(left, right);  // NOLINT
```

Comment:
Why can't we pass the predicate condition in but need to translate from PredicateCondition::Equals to Equals?